### PR TITLE
uses labels instead of OwnerReferences when migrating clusters

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: adjusts the tool to use `labels` instead of `OwnerReferences` when migrating `cluster` resources. It turned out that we cannot use OwnerReferences because there is no referenced object on the seed clusters and the resource will be `gc'ed`.

**Special notes for your reviewer**: see for more https://github.com/kubermatic/kubermatic/pull/1839

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: api/cmd/projects-migrator/main.go [hunk 1]
```diff
@@ -524,18 +566,12 @@ func migrateToProject(ctx migrationContext) error {
...
-					UID:        project.GetUID(),
-					Name:       project.Name,
-				}
 				for _, clusterResource := range clusterResourcesProviderTuple.clusters {
 					if !projectExists {
 						glog.Warningf("skipping cluster resource = %s (physical location %s), there is no project for a user with ID = %s (the user doesn't exist ?)", clusterResource.Name, clusterName, userID)
 						continue
 					}
-					clusterResource.OwnerReferences = append(clusterResource.OwnerReferences, ownerRef)
+					clusterResource.Labels[kubermaticv1.ProjectIDLabelKey] = project.Name
```
Comment (p0lyn0mial) [inline api/cmd/projects-migrator/main.go:570]: uses `label` not `OwneRef` when migrating a cluster.

File: api/cmd/projects-migrator/main.go [hunk 2]
```diff
@@ -587,9 +623,10 @@ type clustersProviderTuple struct {
 }
 
 // getAllClusters gets all clusters resources in the system and groups them by user and cluster (physical location)
-func getAllClusters(ctx migrationContext) (map[string]map[string]*clustersProviderTuple, error) {
+func getAllClusters(ctx migrationContext) (map[string]map[string]*clustersProviderTuple, []*clustersProviderTuple, error) {
```
Comment (p0lyn0mial) [inline api/cmd/projects-migrator/main.go:622]: also returns already migrated clusters (the ones with OwnerRef)

File: api/cmd/projects-migrator/main.go [hunk 3]
```diff
@@ -636,29 +679,37 @@ func getAllClusters(ctx migrationContext) (map[string]map[string]*clustersProvid
...
 		}
 	}
 
-	return clustersToAdoptByUserID, nil
+	return clustersToAdoptByUserID, alreadyAdoptedClusters, nil
 }
 
 // isOwnedByProject is a helper function that extract projectName from the given OwnerReferences
-func isOwnedByProject(owners []metav1.OwnerReference) string {
+func isOwnedByProject(owners []metav1.OwnerReference, labels map[string]string) string {
```
Comment (p0lyn0mial) [inline api/cmd/projects-migrator/main.go:678]: we need to support both approaches the old one (OwnerRef) and the new one (labels) when finding clusters.

File: api/cmd/projects-migrator/main.go [hunk 4]
```diff
@@ -497,7 +539,7 @@ func migrateToProject(ctx migrationContext) error {
 				continue
 			}
 
-			projectName := isOwnedByProject(sshKey.OwnerReferences)
+			projectName := isOwnedByProject(sshKey.OwnerReferences, nil)
```
Comment (p0lyn0mial) [inline api/cmd/projects-migrator/main.go:538]: passing `nil` because we don't use `labels` for ssh keys

File: api/cmd/projects-migrator/main.go [hunk 5]
```diff
@@ -146,6 +141,53 @@ func main() {
 	if err != nil {
 		glog.Errorf("PHASE 4 failed due to %v", err)
 	}
+
+	// phase 5: find clusters that are assigned to a project
+	//          and replace OwnerReferences with a label
+	err = remigrate(ctx)
```
Comment (p0lyn0mial) [inline api/cmd/projects-migrator/main.go:143]: this step will update already migrated clusters (OwnerRef)

File: api/cmd/projects-migrator/main.go [hunk 6]
```diff
@@ -619,15 +656,21 @@ func getAllClusters(ctx migrationContext) (map[string]map[string]*clustersProvid
 	{
 		masterClusters, err := ctx.masterKubermaticClient.KubermaticV1().Clusters().List(metav1.ListOptions{})
```
Comment (mrIncompetent) [inline api/cmd/projects-migrator/main.go]: Is that actually necessary?
Cluster resources only live inside the seed-clusters.
Sure a Seed might be master as well, but then it is already covered by the seed-part, right?

Comment (p0lyn0mial) [reply api/cmd/projects-migrator/main.go]: done.

**Comment by p0lyn0mial** [comment]
@mrIncompetent comments addressed, PTAL

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**